### PR TITLE
create adapter adyoulike

### PR DESCRIPTION
**Pull request by jbAdyoulike**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: 'adyoulike',
  params: {
     placement: "194f787b85c829fb8822cdaf1ae64435",
  }
}
```
-dev-ssp@adyoulike.com
- [x] official adapter submission

### CODE REVIEW

File: src/adapters/adyoulike.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+import * as utils from 'src/utils';
+import { format } from 'src/url';
+import { ajax } from 'src/ajax';
+import { STATUS } from 'src/constants';
+
+var AdyoulikeAdapter = function AdyoulikeAdapter() {
+  const _VERSION = "0.1";
+
+  const baseAdapter = Adapter.createNew('adyoulike');
+  const bidRequests = {};
```
Comment (dbemiller) [inline src/adapters/adyoulike.js]: Keeping state out here will be buggy. What would happen if `callBids` gets called twice before your ajax request returns? For example, suppose a page makes more than one cal to `requestBids`.

You can make it safe if you use `newResponseHandler(requests)` as your ajax callback (where `newResponseHandler(requests)` returns the `handleResponse` function, with the requests stored in the closure scope)

**Comment by dbemiller** [comment]
I'm getting the following error when testing:

`XMLHttpRequest cannot load http://hb-api.omnitagjs.com/hb-api/prebid?. Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. Origin 'http://test.localhost:9999' is therefore not allowed access. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.`

The test page I used (run from the integrationExamples folder) is below:

```
<html>
<head>
    <!-- Prebid Config Section START -->
    <!-- Make sure this is inserted before your GPT tag -->
    <script>
      var PREBID_TIMEOUT = 700;

      var adUnits = [{
        code: 'div-gpt-ad-1460505748561-0',
        sizes: [[300, 250], [300,600]],
        bids: [{
          bidder: 'adyoulike',
          params: {
            placement: "d135a2027b1041059f06fbf8cf4ddfef",
          }
        }]
      }];

      var pbjs = pbjs || {};
      pbjs.que = pbjs.que || [];

    </script>
    <!-- Prebid Config Section END -->

    <!-- Prebid Boilerplate Section START. No Need to Edit. -->
    <script type="text/javascript" src="../build/dev/prebid.js" async></script>
    <script>
      var googletag = googletag || {};
      googletag.cmd = googletag.cmd || [];
      googletag.cmd.push(function() {
        googletag.pubads().disableInitialLoad();
      });

      pbjs.que.push(function() {
        pbjs.addAdUnits(adUnits);
        pbjs.requestBids({
          bidsBackHandler: sendAdserverRequest
        });
      });

      function sendAdserverRequest() {
        if (pbjs.adserverRequestSent) return;
        pbjs.adserverRequestSent = true;
        googletag.cmd.push(function() {
          pbjs.que.push(function() {
            pbjs.setTargetingForGPTAsync();
            googletag.pubads().refresh();
          });
        });
      }

      setTimeout(function() {
        sendAdserverRequest();
      }, PREBID_TIMEOUT);

    </script>
    <!-- Prebid Boilerplate Section END -->

    <script>
      (function () {
        var gads = document.createElement('script');
        gads.async = true;
        gads.type = 'text/javascript';
        var useSSL = 'https:' == document.location.protocol;
        gads.src = (useSSL ? 'https:' : 'http:') +
          '//www.googletagservices.com/tag/js/gpt.js';
        var node = document.getElementsByTagName('script')[0];
        node.parentNode.insertBefore(gads, node);
      })();
    </script>

    <script>
      googletag.cmd.push(function () {
        googletag.defineSlot('/19968336/header-bid-tag-0', [[300, 250], [300, 600]], 'div-gpt-ad-1460505748561-0').addService(googletag.pubads());

        googletag.pubads().enableSingleRequest();
        googletag.enableServices();
      });
    </script>
</head>

<body>
<h2>Prebid.js Test</h2>
<h5>Div-1</h5>
<div id='div-gpt-ad-1460505748561-0'>
    <script type='text/javascript'>
      googletag.cmd.push(function() { googletag.display('div-gpt-ad-1460505748561-0'); });
    </script>
</div>
</body>
</html>
```

**Comment by dbemiller** [comment]
You'll also want to submit a [documentation PR](https://github.com/prebid/prebid.github.io) for the [bidder params page](http://prebid.org/dev-docs/bidders.html).

**Comment by dbemiller** [comment]
I'm seeing CORS errors when testing:

```
XMLHttpRequest cannot load http://hb-api.omnitagjs.com/hb-api/prebid?. Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. Origin 'http://test.localhost:9999' is therefore not allowed access. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.
```

My test page is below:

```
<html>
<head>
    <!-- Prebid Config Section START -->
    <!-- Make sure this is inserted before your GPT tag -->
    <script>
      var PREBID_TIMEOUT = 700;

      var adUnits = [{
        code: 'div-gpt-ad-1460505748561-0',
        sizes: [[300, 250], [300,600]],
        bids: [{
          bidder: 'adyoulike',
          params: {
            placement: "d135a2027b1041059f06fbf8cf4ddfef",
          }
        }]
      }];

      var pbjs = pbjs || {};
      pbjs.que = pbjs.que || [];

    </script>
    <!-- Prebid Config Section END -->

    <!-- Prebid Boilerplate Section START. No Need to Edit. -->
    <script type="text/javascript" src="../build/dev/prebid.js" async></script>
    <script>
      var googletag = googletag || {};
      googletag.cmd = googletag.cmd || [];
      googletag.cmd.push(function() {
        googletag.pubads().disableInitialLoad();
      });

      pbjs.que.push(function() {
        pbjs.addAdUnits(adUnits);
        pbjs.requestBids({
          bidsBackHandler: sendAdserverRequest
        });
      });

      function sendAdserverRequest() {
        if (pbjs.adserverRequestSent) return;
        pbjs.adserverRequestSent = true;
        googletag.cmd.push(function() {
          pbjs.que.push(function() {
            pbjs.setTargetingForGPTAsync();
            googletag.pubads().refresh();
          });
        });
      }

      setTimeout(function() {
        sendAdserverRequest();
      }, PREBID_TIMEOUT);

    </script>
    <!-- Prebid Boilerplate Section END -->

    <script>
      (function () {
        var gads = document.createElement('script');
        gads.async = true;
        gads.type = 'text/javascript';
        var useSSL = 'https:' == document.location.protocol;
        gads.src = (useSSL ? 'https:' : 'http:') +
          '//www.googletagservices.com/tag/js/gpt.js';
        var node = document.getElementsByTagName('script')[0];
        node.parentNode.insertBefore(gads, node);
      })();
    </script>

    <script>
      googletag.cmd.push(function () {
        googletag.defineSlot('/19968336/header-bid-tag-0', [[300, 250], [300, 600]], 'div-gpt-ad-1460505748561-0').addService(googletag.pubads());

        googletag.pubads().enableSingleRequest();
        googletag.enableServices();
      });
    </script>
</head>

<body>
<h2>Prebid.js Test</h2>
<h5>Div-1</h5>
<div id='div-gpt-ad-1460505748561-0'>
    <script type='text/javascript'>
      googletag.cmd.push(function() { googletag.display('div-gpt-ad-1460505748561-0'); });
    </script>
</div>
</body>
</html>
```

**Comment by jbAdyoulike** [comment]
Could you use this placement ID for your test: `194f787b85c829fb8822cdaf1ae64435` ?

**Comment by dbemiller** [comment]
@jbAdyoulike please fix the linting errors

**Comment by jbAdyoulike** [comment]
@dbemiller done

**Comment by dbemiller** [comment]
Am seeing errors using that placement Id:
screen shot 2017-05-19 at 3 53 11 pm

Request params:

```
      var adUnits = [{
        code: 'div-gpt-ad-1460505748561-0',
        sizes: [[300, 250], [300,600]],
        bids: [{
          bidder: 'adyoulike',
          params: {
            placement: "194f787b85c829fb8822cdaf1ae64435",
          }
        }]
      }];
```

**Comment by jbAdyoulike** [comment]
@dbemiller Could you give me the request body and response body of this endpoint `http://hb-api.omnitagjs.com/hb-api/prebid` from the network console to let me investigate on my side ? 

This error looks strange, Following what I have from my side :
cap-prebid

**Comment by dbemiller** [comment]
screen shot 2017-05-22 at 9 43 25 am
screen shot 2017-05-22 at 9 43 40 am

**Comment by jbAdyoulike** [comment]
@dbemiller Ok, It was a misconfiguration on my placement, it should works now.

File: src/adapters/adyoulike.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    }
+    return true;
+  }
+
+  /* Get current page referrer url */
+  function getReferrerUrl() {
+    try {
+      return window.top.document.referrer;
+    } catch (e) { }
+    return document.referrer;
```
Comment (dbemiller) [inline src/adapters/adyoulike.js]: Generally, it's frowned upon to use try/catch for control flow: https://softwareengineering.stackexchange.com/a/189225

You can write this linearly by testing for undefined (the same goes for `getCanonicalUrl`).

File: src/adapters/adyoulike.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  /* Get current page canonical url */
+  function getCanonicalUrl() {
+    if (utils.inIframes) {
```
Comment (dbemiller) [inline src/adapters/adyoulike.js]: This is undefined... did you mean to call `utils.inIframe()` here?

**Comment by dbemiller** [comment]
Needs just a few changes as far as I can tell... but I do want to complement you especially for writing such pretty code :). Super clear and easy to read.

**Comment by jbAdyoulike** [comment]
@dbemiller It's a pleasure ;)

I have created the PR for the bidder params page prebid/prebid.github.io#252

**Comment by ayl-julien** [comment]
Hi @protonate what do we have to do to have our name updated on prebid.org?

**Comment by dbemiller** [comment]
@ayl-julien @protonate is away for a little while, but submit a PR against [this repo](https://github.com/prebid/prebid.github.io) to change it.

**Pull request merged**